crates/edr_napi/src/solidity_tests.rs

```diff
@@ -55,15 +64,35 @@ pub fn run_solidity_tests(
         .map(TryInto::try_into)
         .collect::<Result<Vec<_>, _>>()?;
 
-    let runner = build_runner(&known_contracts, test_suites, gas_report)?;
-
     let (tx_results, mut rx_results) = tokio::sync::mpsc::unbounded_channel::<(
         foundry_common::ArtifactId,
         forge::result::SuiteResult,
     )>();
 
     let runtime = runtime::Handle::current();
     runtime.spawn(async move {
+        let runner = match build_runner(&known_contracts, test_suites, config_args).await {
+            Ok(runner) => runner,
+            Err(error) => {
+                let call_status =
+                    error_callback_fn.call(error, ThreadsafeFunctionCallMode::Blocking);
+                // This should always succeed since we're using an unbounded queue. We add an
+                // assertion for completeness.
+                assert!(
+                    matches!(call_status, napi::Status::Ok),
+                    "Failed to call callback with status {call_status:?}"
+                );
+                return;
+            }
+        };
```

Comment:
> The reason for introducing this error callback is that building a test runner is async now to resolve forked EVM parameters, but the JS interface is sync and returns results with a progress callback. If we want to be able to propagate errors encountered while building a test runner we have two options:
>
> 1. Make the test runner async. This is possible, but we'd need to create a test runner class that takes the progress callback as argument first and then give it an async method for reasons explained in https://github.com/NomicFoundation/edr/pull/563.

I don't understand how the problem described in the other pull request applies here. To me it seems like a different use case, which wouldn't suffer from the same problem as describe in PR #563.

We have many examples of `async fn` in our N-API that have `.await` calls that work just fine. Can't this function also be made async for the construction of the build runner and only afterwards spawn the background process for parallelisation - which does suffer from the issue in PR #563?

That seems more idiomatic than the error callback.

Am I overlooking something?

Reply:
The issue is that `run_solidity_tests` takes a progress callback argument as a `napi::JsFunction` and `napi::JsFunction` is `!Send`, so it cannot be passed into an async function as argument on the Rust side. 

So our options are either working around this by introducing a test runner class that takes the progress callback as argument in a sync constructor and exposes an async method to start running the tests that'd throw if there is an error when constructing the test runner. Or alternatively add an error callback argument which is quite common in NodeJS callback-based interfaces. Exposing an error callback lead to simpler code on both sides of the FFI, so I went with that.

Reply:
It's not self-evident but N-API provides a mechanism to work around the `!Send` issue. We use it for [the constructor of the `Provider`](https://github.com/NomicFoundation/edr/blob/main/crates/edr_napi/src/provider.rs#L32), which is also `async`.

To validate that my idea actually works, I implemented it [here](https://github.com/NomicFoundation/edr/pull/600).

The `run_solidity_tests` function is a function that "spawns a Solidity test runner". In line with the rest of our API, this change ensures that the function will either spawn the test runner or throw an exception. All of the additional progress then works with the callback.

Reply:
Thanks that's good to know! I'm on mobile, could you also update the js test in edr_napi in your PR to see how that'd look like?

Reply:
Done https://github.com/NomicFoundation/edr/pull/600/commits/6d597cf2c604dc965d3529828496270a3569cffd

Reply:
Thanks! @fvictorio could you review the JS changes? Is the promise going to be rejected if the async executor throws?

Reply:
I don't think I can answer this just by looking at the changes in that commit, since it only has changes in the `.d.ts` file.

The way javascript works here is a bit weird. If you have this code:

```js
function foo() {
  throw new Error("foo")
  return Promise.resolve(42);
}

async function main() {
  const answer = foo().catch(() => {});
}

main()
```

Then `main` will throw, because `foo()` doesn't return a rejected promise. But if you add `async` to `foo`'s declaration:

```js
async function foo() {
  throw new Error("foo")
  return Promise.resolve(42);
}

async function main() {
  const answer = foo().catch(() => {});
}

main()
```

then the error will be converted into a rejected promise, and `main` won't throw. So whether or not the function will return a rejected promise or throw an error is hard to answer without knowing what N-API does under the hood. The fact that there weren't any changes in the `index.js` file makes me think that it _might_ throw (but, again, hard to know; it might as well return a rejected promise).

But this should be relatively easy to manually test to see what's the behavior, right?

Reply:
Thanks, I added a test for it in https://github.com/NomicFoundation/edr/commit/94ca87a63a0b482630b9d9446d3f7691dfac50ac and it looks like errors aren't thrown with the async executor unfortunately. Using a sync executor and passing `reject` to `catch` works, but that's just a more complicated way of passing an error callback, so I'd respectfully decline this change request.

Reply:
> error aren't thrown with the async executor

Do you mean that the function doesn't throw an error but the promise rejects, the opposite, or that neither happens?

In any case, I'm ok with keeping the error callback approach, following previous decisions about not caring too much about the API here as long as Hardhat is the only consumer, and the fact that at some point we'll offer a wrapper with a nicer API unconstrained by `napi-rs` limitations.

Reply:
> Do you mean that the function doesn't throw an error but the promise rejects, the opposite, or that neither happens?

The promise just hangs if there is an error: https://github.com/NomicFoundation/edr/actions/runs/10416449901/job/28849161857#step:8:42

And this is how the napi function is invoked: https://github.com/NomicFoundation/edr/commit/94ca87a63a0b482630b9d9446d3f7691dfac50ac#diff-ed50e3e633d981c1abfaec950d3a22878e0f1a4ee8a83722fd1d86f5bb001616R13

> In any case, I'm ok with keeping the error callback approach, following previous decisions about not caring too much about the API here as long as Hardhat is the only consumer, and the fact that at some point we'll offer a wrapper with a nicer API unconstrained by napi-rs limitations.

Ok cheers

Reply:
> Thanks, I added a test for it in https://github.com/NomicFoundation/edr/commit/94ca87a63a0b482630b9d9446d3f7691dfac50ac and it looks like errors aren't thrown with the async executor unfortunately.

That's concerning for all of our other async code though, isn't it?

I'm curious to try one thing. In this code I'm explicitly rejecting the `Promise` instead of resolving it with an error. I'll test whether that makes a difference, because otherwise all of our async code might be at risk of this problem.

Reply:
So I tried a `try..catch` block around the `await runSolidityTests` and logged the error. It was in fact correctly returning the error.

The way that `new Promise` works is that unless you are explicitly forwarding the thrown error to the newly defined promises `reject` function, it will not reject the outer Promise when an inner `Promise` is rejected. In the sync code, you are explicitly doing this using the `.catch(reject)`. Adding that to the async code ensures that the promise is forwarded.

If you want to use `await runSolidityTests`, you have to define an `async fn` which does the piping of inner `Promise` rejections for you.

I found these two articles helpful:
- https://javascript.info/promise-error-handling#unhandled-rejections
- https://stackoverflow.com/questions/45348955/using-await-within-a-promise

Reply:
> That's concerning for all of our other async code though, isn't it?

Normally you wouldn't wrap a JS function that returns a promise in another promise. The issue here is that you can pass an async executor to a  `Promise` and call `await runSolidityTests(...)` in the executor of the promise, but if that call throws, the `Promise` won't be rejected. 

I think it's quite clear now that making `runSolidityTests` async is not a good idea, because it doesn't compose well with its usage. Could you approve please so that I can merge it and then you can run experiments as a separate issue?

Reply:
> I think it's quite clear now that making runSolidityTests async is not a good idea, because it doesn't compose well with its usage.

In my [previous comment](https://github.com/NomicFoundation/edr/pull/592#discussion_r1719777489) I explained why it was failing. It wasn't a result of `runSolidityTests` being async, but instead an error in the piping of the promises rejection; something that's part of Javascript's design of the `Promise` API.